src/primitives.rs

```diff
@@ -280,9 +287,73 @@ where
     }
 }
 
-impl<T> Versionize for Vec<T>
+macro_rules! impl_versionize_vec_like_type {
+    ($VecType:ident <$GenericParam:ident>) => {
+        impl<$GenericParam> Versionize for $VecType<$GenericParam>
+        where
+            T: Versionize,
+        {
+            #[inline]
+            fn serialize<W: std::io::Write>(
+                &self,
+                mut writer: &mut W,
+                version_map: &VersionMap,
+                app_version: u16,
+            ) -> VersionizeResult<()> {
+                if self.len() > MAX_VEC_SIZE / std::mem::size_of::<T>() {
+                    return Err(VersionizeError::VecLength(self.len()));
+                }
+
+                // Serialize in the same fashion as bincode:
+                // Write len.
+                bincode::serialize_into(&mut writer, &self.len())
+                    .map_err(|ref err| VersionizeError::Serialize(format!("{:?}", err)))?;
+                // Walk the vec and write each element.
+                for element in self {
+                    element.serialize(writer, version_map, app_version)?;
+                }
+                Ok(())
+            }
+
+            #[inline]
+            fn deserialize<R: std::io::Read>(
+                mut reader: &mut R,
+                version_map: &VersionMap,
+                app_version: u16,
+            ) -> VersionizeResult<Self> {
+                let len: usize = bincode::deserialize_from(&mut reader)
+                    .map_err(|ref err| VersionizeError::Deserialize(format!("{:?}", err)))?;
+
+                if len > MAX_VEC_SIZE / std::mem::size_of::<T>() {
+                    return Err(VersionizeError::VecLength(len));
+                }
+
+                let mut v = Vec::with_capacity(len);
+
+                for _ in 0..len {
+                    let element: T = T::deserialize(reader, version_map, app_version)
+                        .map_err(|ref err| VersionizeError::Deserialize(format!("{:?}", err)))?;
+                    v.push(element);
+                }
+                Ok(v.into())
+            }
+
+            // Not used yet.
+            fn version() -> u16 {
+                1
+            }
+        }
+    };
+}
+
+impl_versionize_vec_like_type!(Vec<T>);
+impl_versionize_vec_like_type!(VecDeque<T>);
```

Comment:
I haven't looked at the implementation details, but from a design point of view the idea was to use versionize only on state to reduce the need of bumping versions when some implementation details change, but the state would stay the same.

In this case `VecDequeue` is a `Vec` that also has some implementation details assigned to it (i.e. how the pop operation looks). If we would look at the pure state of a `VecDequeue` it would still be a vector. In this case it would be better to simply save the `Vec` as the state. If you change the implementation details in your VMM and you no longer want to use `VecDequeue` and maybe you want to use `Vec` instead, there will be no version bump needed because your state stays the same and only the implementation changes.

Reply:
Sure, in our scenario, we may implement Versionize for some state which not only used to save/restore, but used when program is running. This design would simplify some code of hypervisor, because we don't need to convert runtime state and snapshot state by hand when doing snapshot/restore.